// @axelar-network/axelar-local-dev/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.forkAndExport = exports.createAndExport = exports.relay = exports.mainnetInfo = exports.testnetInfo = exports.networks = exports.getDepositAddress = exports.stopAll = exports.stop = exports.setupNetwork = exports.getAllNetworks = exports.getNetwork = exports.forkNetwork = exports.createNetwork = exports.listen = exports.getGasPrice = exports.getFee = exports.utils = void 0;
var exportUtils_1 = require("./exportUtils");
Object.defineProperty(exports, "createAndExport", { enumerable: true, get: function () { return exportUtils_1.createAndExport; } });
Object.defineProperty(exports, "forkAndExport", { enumerable: true, get: function () { return exportUtils_1.forkAndExport; } });
var relay_1 = require("./relay");
Object.defineProperty(exports, "relay", { enumerable: true, get: function () { return relay_1.relay; } });
var utils_1 = require("./utils");
var info_1 = require("./info");
Object.defineProperty(exports, "testnetInfo", { enumerable: true, get: function () { return info_1.testnetInfo; } });
Object.defineProperty(exports, "mainnetInfo", { enumerable: true, get: function () { return info_1.mainnetInfo; } });
var Network_1 = require("./Network");
Object.defineProperty(exports, "networks", { enumerable: true, get: function () { return Network_1.networks; } });
var networkUtils_1 = require("./networkUtils");
Object.defineProperty(exports, "getFee", { enumerable: true, get: function () { return networkUtils_1.getFee; } });
Object.defineProperty(exports, "getGasPrice", { enumerable: true, get: function () { return networkUtils_1.getGasPrice; } });
Object.defineProperty(exports, "listen", { enumerable: true, get: function () { return networkUtils_1.listen; } });
Object.defineProperty(exports, "createNetwork", { enumerable: true, get: function () { return networkUtils_1.createNetwork; } });
Object.defineProperty(exports, "forkNetwork", { enumerable: true, get: function () { return networkUtils_1.forkNetwork; } });
Object.defineProperty(exports, "getNetwork", { enumerable: true, get: function () { return networkUtils_1.getNetwork; } });
Object.defineProperty(exports, "getAllNetworks", { enumerable: true, get: function () { return networkUtils_1.getAllNetworks; } });
Object.defineProperty(exports, "setupNetwork", { enumerable: true, get: function () { return networkUtils_1.setupNetwork; } });
Object.defineProperty(exports, "stop", { enumerable: true, get: function () { return networkUtils_1.stop; } });
Object.defineProperty(exports, "stopAll", { enumerable: true, get: function () { return networkUtils_1.stopAll; } });
Object.defineProperty(exports, "getDepositAddress", { enumerable: true, get: function () { return networkUtils_1.getDepositAddress; } });
exports.utils = {
    deployContract: utils_1.deployContract,
    defaultAccounts: utils_1.defaultAccounts,
    setJSON: utils_1.setJSON,
    setLogger: utils_1.setLogger,
};
